packages/i18n/src/create-i18n.js

```diff
@@ -147,24 +155,161 @@ export const createI18n = ( initialData, initialDomain ) => {
 		return tannin.dcnpgettext( domain, context, single, plural, number );
 	};
 
+	/** @type {GetFilterDomain} */
+	const getFilterDomain = ( domain ) => {
+		if ( typeof domain === 'undefined' ) {
+			return 'default';
+		}
+		return domain;
+	};
+
 	/** @type {__} */
 	const __ = ( text, domain ) => {
-		return dcnpgettext( domain, undefined, text );
+		let translation = dcnpgettext( domain, undefined, text );
+		/**
+		 * Filters text with its translation.
+		 *
+		 * @param {string} translation Translated text.
+		 * @param {string} text        Text to translate.
+		 * @param {string} domain      Text domain. Unique identifier for retrieving translated strings.
+		 */
+		if ( typeof hooks === 'undefined' ) {
+			return translation;
+		}
+		translation = String(
```

Comment:
Converting the translation to always be of a string type seems a bit restrictive. Is there a reason not to allow it to be filtered into any type of value?

Reply:
The type definitions / JSDoc say that it returns a string, so it makes sense to uphold that promise, no?

Reply:
Would it be worth removing the type conversion and updating the JSDoc? I think it will make the filters a lot more flexible, especially in a React context, where it might not be uncommon to need to return a React component directly as a result of the `gettext` call.

Reply:
When used in a React component:
```jsx
<span>{ __( 'Hello' ) }</span>
```
The return value of `__()` can be a generic `ReactNode`, which includes many possible value types, `string` being one of them.

It would be nice to keep this flexibility -- there are creative use cases that benefit from it.

The return type of `hooks.applyFilters` is `unknown`. If we want to coerce that into a `string`, we can use a type annotation:
```js
translation = /** @type {string} */ hooks.applyFilters();
```
That affects only the type checker, and doesn't do runtime conversions like `String()` does.

Reply:
I'm more than happy to make this change (The String() calls always made me feel a little uneasy! I've tried the annotation approach required here, but can't seem to make it compile, and unfortunately IU don't know enough about the syntax to coax it into submission. 

```
packages/i18n/src/create-i18n.js:179:3 - error TS2322: Type 'unknown' is not assignable to type 'string'.

179  	translation = /** @type {string} */ hooks.applyFilters(
     	~~~~~~~~~~~
```

Reply:
@jsnajdr It looks like `applyFilters` returns `unknown` rather than `any` which means (I think) that we can't just type hint it?

Reply:
Interesting. The `unknown` type should be assignable to anything else, without constraints. I though that `value as string` is exactly the same thing as `/** @type {string} */ value`, but TypeScript playground reports error only for the second (jsdoc) one:

<img width="832" alt="Screenshot 2021-01-28 at 10 58 45" src="https://user-images.githubusercontent.com/664258/106121520-f1b2ad80-6157-11eb-991e-79644c62800c.png">

@sirreal Do you have any insight about what's going on? Maybe there is some TS option that affects this?

Reply:
@leewillis77 I found that this workaround works:
```js
translation = /** @type {string} */ ( /** @type {*} */ applyFilters( ... ) );
```

Reply:
@jsnajdr Thanks. PR updated.

Reply:
> Interesting. The `unknown` type should be assignable to anything else, without constraints.
> 
> @sirreal Do you have any insight about what's going on? Maybe there is some TS option that affects this?

Anything can be assigned _to_ the `unknown` type, but `unknown` does not satisfy any other types. `any` is the type that satisfies all other type (and anything can be assigned to `any`.

You can read more about [`unknown` in its announcement](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-0.html#new-unknown-top-type).

> TypeScript 3.0 introduces a new top type unknown. unknown is the type-safe counterpart of any. Anything is assignable to unknown, but unknown isn’t assignable to anything but itself and any without a type assertion or a control flow based narrowing. Likewise, no operations are permitted on an unknown without first asserting or narrowing to a more specific type.

When we really don't have type information, especially at the boundaries of our typed application like a REST response or receiving data across untyped APIs like here, `unknown` makes that very clear and forces you to narrow with guards or type assertions. Before `unknown`, there was only `any` which was completely unsafe because you can put anything in and do anything with it.

The error is expected and the fix is what TypeScript recommends (this is a type assertion written in JSDoc). First, we assert the `unknown` is `any`, then we can narrow it to the type of our choosing because `any`:

```js
translation = /** @type {string} */ ( /** @type {*} */ applyFilters( ... ) );
```

A (safer) alternative would be to use a [type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#typeof-type-guards), but has a runtime cost:

```js
const filteredTranslation = applyFilters( /* … */ );
return typeof filteredTranslation === 'string' ? filteredTranslation : translation; // we know it's a string
```